---
title:  "Antigen Tracking Figures"
author: "Ryan Sheridan"
date:   "`r Sys.Date()`"

output: 
  pdf_document:
    latex_engine: xelatex
    fig_caption: false

documentclass: article
mainfont: DejaVu Sans
geometry: top=1in,bottom=1in,left=1in,right=1in

params:
  template_dir: "src"
  rds_dir:      "sobjs/so"
  ref_dir:      "ref"
  xlsx_dir:     "results/figures"
  browser_dir:  "results/cellbrowser"
  
  # These parameters specify the input matrices and cell types that should be used for analysis
  # mat_in: Cellranger output folder.
  # cell_type: Cell type used for subsetting Seurat objects.
  # title: Section title.
  # sobj_out: The output Seurat object that is saved after processing and subsetting based on cell type.
  sobjs:
    value:
      d2_DC:
        mat_in:    "results/GEX_CD45pos_d2-ADT_CD45pos_d2/outs/filtered_feature_bc_matrix"
        cell_type: "DC"
        title:     "Myeloid Cells (Day 2)"
        sobj_out:  "d2_DC_so_2.rds"
      d14_DC:
        mat_in:    "results/GEX_CD45pos_d14-ADT_CD45pos_d14/outs/filtered_feature_bc_matrix"
        cell_type: "DC"
        title:     "Myeloid Cells (Day 14)"
        sobj_out:  "d14_DC_so_2.rds"
      d2_LEC:
        mat_in:    "results/GEX_CD45neg_d2-ADT_CD45neg_d2/outs/filtered_feature_bc_matrix"
        cell_type: "LEC"
        title:     "Endothelial/Stromal Cells (Day 2)"
        sobj_out:  "d2_LEC_so_2.rds"
      d14_LEC:
        mat_in:    "results/GEX_CD45neg_d14-ADT_CD45neg_d14/outs/filtered_feature_bc_matrix"
        cell_type: "LEC"
        title:     "Endothelial/Stromal Cells (Day 14)"
        sobj_out:  "d14_LEC_so_2.rds"
      d2_FRC:
        mat_in:    "results/GEX_CD45neg_d2-ADT_CD45neg_d2/outs/filtered_feature_bc_matrix"
        cell_type: "fibroblast"
        title:     "Fibroblast/Stromal Cells (Day 2)"
        sobj_out:  "d2_FRC_so_2.rds"
      d14_FRC:
        mat_in:    "results/GEX_CD45neg_d14-ADT_CD45neg_d14/outs/filtered_feature_bc_matrix"
        cell_type: "fibroblast"
        title:     "Fibroblast/Stromal Cells (Day 14)"
        sobj_out:  "d14_FRC_so_2.rds"
        
  # These parameters specify paths to clustifyr references
  ref_sobjs: ["d2_so.rds", "d14_so.rds"]
  type_ref:  "ref_celltype_walsh.rda"
  xiang_ref: "ref_LEC_xiang.rda"
  subtype_refs:
    value:
      DC:         "ref_DC_walsh.rda"
      LEC:        "ref_LEC_walsh.rda"
      fibroblast: "ref_FRC_walsh.rda"
---

```{r "Chunk opts", echo = F}

# Default chunk options
knitr::opts_chunk$set(
  message = F,
  warning = F,
  echo    = F
)

```

`r knitr::knit_child(here::here(params$template_dir, "setup.Rmd"))`

```{r "Plotting functions", echo = F}

# Set equal x-axis scales
equalize_x <- function(gg_list_in, log_tran = T, ...) {
  
  set_lims <- function(gg_in, min_x, max_x, log_tran, ...) {
    res <- gg_in +
      coord_cartesian(xlim = c(min_x, max_x))
    
    if (log_tran) {
      res <- res +
        scale_x_log10(labels = trans_format("log10", math_format(10^.x)), ...)
    }
    
    res
  }
  
  gg_ranges <- gg_list_in %>%
    map(~ ggplot_build(.x)$layout$panel_scales_x[[1]]$range$range)
  
  min_val <- gg_ranges %>%
    map_dbl(~ .x[1]) %>%
    min()
  
  max_val <- gg_ranges %>%
    map_dbl(~ .x[2]) %>%
    max()
  
  res <- gg_list_in %>%
    map(
      set_lims,
      min_x    = min_val, 
      max_x    = max_val, 
      log_tran = log_tran,
      breaks   = 10^(-4:4),
      ...
    )
  
  res
}

# Create figure 3 panels
create_fig3 <- function(sobj_in, cols_in, subtype_column = "subtype", data_slot = "counts", ova_cols = c("#fafafa", "#d7301f"),
                        box_counts = c("Relative ova signal" = "ova_fc"), umap_counts = c("ova counts" = "adt_ovalbumin"), 
                        plot_title = NULL, pt_size = 0.1, pt_outline = 0.4, pt_size_2 = 0.3, pt_outline_2 = 0.5, box_cell_count = T,
                        control_types = c("B cell", "T cell"), on_top = NULL, ...) {
  
  box_column <- umap_column <- "cell_type"
  box_cols <- umap_cols <- cols_in
  
  # Fetch plotting data
  data_df <- sobj_in %>%
    FetchData(c(subtype_column, box_counts, umap_counts, "UMAP_1", "UMAP_2"), slot = data_slot) %>%
    as_tibble(rownames = "cell_id")
  
  if (!is.null(names(box_counts))) {
    data_df <- data_df %>%
      rename(!!box_counts)
    
    box_counts <- names(box_counts)
  }
  
  if (!is.null(names(umap_counts))) {
    data_df <- data_df %>%
      rename(!!umap_counts)
    
    umap_counts <- names(umap_counts)
  }
  
  # Set subtype order
  # Move select cell types to front of order
  data_df <- data_df %>%
    mutate(
      cell_type = !!sym(subtype_column),
      cell_type = fct_reorder(cell_type, !!sym(box_counts), median)
    )
  
  type_order <- levels(data_df$cell_type)
  
  if (!is.null(control_types)) {
    control_types <- control_types[control_types %in% type_order]
    type_order <- type_order[!type_order %in% control_types]
    type_order <- c(control_types, type_order)
  }
  
  # Count cells for each subtype
  data_df <- data_df %>%
    group_by(cell_type) %>%
    mutate(cell_count = n_distinct(cell_id)) %>%
    ungroup() %>%
    mutate(cell_type = fct_relevel(cell_type, type_order)) %>%
    arrange(cell_type) %>%
    mutate(
      cell_count = str_c(cell_type, "\n(n = ", cell_count, ")"),
      cell_count = fct_inorder(cell_count)
    )
  
  # Set cell type colors
  names(type_order) <- levels(data_df$cell_count)
  
  cols_df <- tibble(
    cell_type = type_order,
    cell_count = names(type_order)
  )
  
  cols_df <- cols_df %>%
    mutate(color = cols_in[cell_type])
  
  # Subtype UMAP
  # Reorder layers of UMAP
  if (!is.null(on_top)) {
    on_top <- umap_cols[on_top]
    
    umap_cols <- umap_cols[!names(umap_cols) %in% names(on_top)]
    umap_cols <- c(umap_cols, on_top)
  }
  
  umap <- data_df %>%
    plot_features(
      feature     = umap_column,
      pt_size     = pt_size,
      pt_outline  = pt_outline,
      plot_cols   = umap_cols,
      feat_levels = names(umap_cols)
    ) +
    guides(color = guide_legend(override.aes = list(size = 3.5))) +
    ggtitle(plot_title) +
    blank_theme +
    theme(
      plot.title = element_text(size = 12),
      legend.position = "none"
    ) +
    theme(...)
  
  # OVA UMAP
  if (!is.null(umap_counts)) {
    ova_umap <- data_df %>%
      plot_features(
        feature    = umap_counts,
        plot_cols  = ova_cols,
        pt_size    = pt_size_2,
        pt_outline = pt_outline_2,
        min_pct    = 0.01,
        max_pct    = 0.99
      ) +
      guides(color = guide_colorbar(frame.colour = "black", frame.linewidth = 0.2)) +
      blank_theme +
      theme(
        plot.title        = element_text(size = 10, hjust = 0.5),
        legend.position   = "right",
        legend.key.width  = unit(0.15, "cm"),
        legend.key.height = unit(0.30, "cm"),
        legend.title      = element_text(size = 10),
        legend.text       = element_text(size = 8)
      )
  }
  
  # OVA boxes
  if (box_cell_count) {
    box_column <- "cell_count"
    box_cols <- set_names(
      x  = cols_df$color,
      nm = cols_df$cell_count
    )
  }
  
  boxes <- data_df %>%
    ggplot(aes(!!sym(box_counts), !!sym(box_column), fill = !!sym(box_column))) +
    geom_violin(size = 0.3, draw_quantiles = c(0.25, 0.75), alpha = 0.75) +
    stat_summary(geom = "point", color = "black", fun = median) +
    
    scale_color_manual(values = box_cols) +
    scale_fill_manual(values = box_cols) +
    theme_minimal_vgrid() +
    theme(
      legend.position    = "none",
      axis.title.y       = element_blank(),
      axis.title         = element_text(size = 10),
      axis.text          = element_text(size = 8),
      axis.ticks.x       = element_line(size = 0.1),
      panel.grid.major.x = element_line(size = 0.1)
    )
  
  res <- list(umap, boxes)
  
  if (!is.null(umap_counts)) {
    res <- append(res, list(ova_umap))
  }
  
  names(res) <- rep(plot_title, length(res))
  
  res
}

# Create GMM panels
create_gmm_panels <- function(sobj_in, gmm_column = "GMM_grp", ova_cols, bar_cols, pt_size = 0.00001,
                              pt_outline = 0.4, show_bars = T, plot_labs = c(letters[1:3], ""), ...) {
  
  # Theme elements
  legd_guide <- guide_legend(override.aes = list(
    size    = 3.5,
    shape   = 21,
    color   = "black",
    stroke  = 0.25,
    reverse = T
  ))
  
  text_theme <- theme(
    axis.title  = element_text(size = 10),
    legend.text = element_text(size = 8),
    axis.text   = element_text(size = 8)
  )
  
  # Data for OVA group UMAP
  ova_order <- names(ova_cols)
  
  data_df <- sobj_in@meta.data %>%
    as_tibble(rownames = "cell_id") %>%
    group_by(!!sym(gmm_column)) %>%
    mutate(cell_count = n_distinct(cell_id)) %>%
    ungroup() %>%
    mutate(!!sym(gmm_column) := fct_relevel(!!sym(gmm_column), ova_order)) %>%
    arrange(!!sym(gmm_column)) %>%
    mutate(
      cell_count = str_c(!!sym(gmm_column), "\n(n = ", cell_count, ")"),
      cell_count = fct_inorder(cell_count)
    )
  
  # Set OVA group colors
  names(ova_order) <- levels(data_df$cell_count)
  
  cols_df <- tibble(
    grp        = ova_order,
    cell_count = names(ova_order)
  )
  
  cols_df <- cols_df %>%
    mutate(color = ova_cols[grp])
  
  umap_cols <- set_names(
    x  = cols_df$color,
    nm = cols_df$cell_count
  )
  
  # Create OVA group UMAP
  ova_labeller <- function(labels) {
    labels %>%
      map(str_remove, "^[a-zA-Z0-9 \\-\\+]+-")
  }
  
  ova_grp_umap <- data_df %>%
    plot_features(
      feature     = "cell_count",
      pt_size     = pt_size,
      pt_outline  = pt_outline,
      plot_cols   = umap_cols,
      feat_levels = names(ova_order)
    ) +
    scale_color_manual(values = umap_cols, labels = ova_labeller) +
    scale_fill_manual(values = umap_cols, labels = ova_labeller) +
    guides(color = legd_guide, fill = legd_guide) +
    text_theme +
    blank_theme +
    theme(
      plot.margin     = unit(c(0.2, 1, 0.2, 1.5), "cm"),
      legend.position = "top",
      legend.title    = element_blank(),
      legend.text     = element_text(size = 8)
    )
  
  # OVA hist
  # Add +1 pseudo count to adt_ovalbumin since using log scale
  ova_hist_order <- names(ova_cols) %>%
    grep("ova (low|high)", ., value = T)
  
  ova_hist <- sobj_in@meta.data %>%
    filter(!!sym(gmm_column) != "Other") %>%
    group_by(!!sym(gmm_column)) %>%
    mutate(mu = mean(adt_ovalbumin)) %>%
    ungroup() %>%
    mutate(!!sym(gmm_column) := fct_relevel(!!sym(gmm_column), ova_hist_order)) %>%
    
    ggplot(aes(adt_ovalbumin + 1, after_stat(density), fill = !!sym(gmm_column))) +
    stat_density(geom = "point", position = "identity", size = 0, color = "white") +  # Need this for square legend symbols
    geom_density(size = 0.3, alpha = 0.8, show.legend = F) +
    
    guides(fill = guide_legend(override.aes = list(                                   # Need this for square legend symbols
      shape  = 22,
      size   = 3.5,
      stroke = 0.25,
      color  = "black"
    ))) +
    
    geom_vline(aes(xintercept = mu), size = 0.5, linetype = 2, color = "grey35") +
    coord_cartesian(ylim = c(0, 1.7)) +
    scale_fill_manual(values = ova_cols, labels = ova_labeller) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x)), breaks = 10^(-4:4)) +
    labs(x = "ova signal (UMI counts + 1)", y = "Density") +
    
    theme_minimal_hgrid() +
    text_theme +
    theme(
      plot.margin        = unit(c(0.2, 0.5, 0.2, 0.2), "cm"),
      legend.position    = c(0.05, 0.92),
      legend.key.height  = unit(0.15, "cm"),
      legend.title       = element_blank(),
      axis.line.y        = element_line(size = 0.5, color = "grey85"),
      axis.ticks.y       = element_line(size = 0.1),
      panel.grid.major.y = element_line(size = 0.1)
    )
  
  # OVA subtype bar graphs
  type_bar <- sobj_in@meta.data %>%
    as_tibble(rownames = "cell_id") %>%
    filter(!!sym(gmm_column) != "Other") %>%
    mutate(
      subtype = fct_reorder(subtype, cell_id, n_distinct),
      !!sym(gmm_column) := fct_relevel(!!sym(gmm_column), c("ova low", "ova high"))
    ) %>%
    
    ggplot(aes(!!sym(gmm_column), fill = subtype)) +
    stat_count(position = "fill", geom = "point", size = 0, color = "white") +    # Need this for square legend symbols
    geom_bar(position = "fill", size = 0.25, color = "black", show.legend = F) +
    
    guides(fill = guide_legend(ncol = 1, override.aes = list(                     # Need this for square legend symbols
      shape  = 22, 
      size   = 3.5, 
      stroke = 0.25, 
      color  = "black"
    ))) +
    
    scale_fill_manual(values = bar_cols) +
    scale_y_continuous(breaks = c(0, 0.5, 1)) +
    labs(y = "Fraction of cells") +
    theme_minimal_hgrid() +
    text_theme +
    theme(
      plot.margin        = unit(c(0.2, 4, 0.2, 0.2), "cm"),
      legend.title       = element_blank(),
      legend.key.height  = unit(0.35, "cm"),
      axis.title.x       = element_blank(),
      axis.line.y        = element_line(size = 0.5, color = "grey85"),
      axis.ticks.y       = element_line(size = 0.1),
      axis.text.x        = element_text(hjust = c(0.6, 0.4)),
      panel.grid.major.y = element_blank()
    )
  
  # Create final figure
  if (!show_bars) {
    type_bar <- ggplot() +
      theme_void()
  }
  
  res <- plot_grid(
    ova_grp_umap, ova_hist, type_bar,
    labels         = plot_labs,
    label_fontface = "plain",
    label_size     = 18,
    ...
  )
  
  res
}

# Create feature UMAPs
create_feat_umap_panels <- function(sobj_in, feats, ref_cols, pt_size = 0.00001, pt_outline = 0.4, n_row = 2,
                                    low_col = c("white", "white"), high_col_mtplyr = 1, on_top = NULL,
                                    return_list = F, ...) {
  
  # Theme elements
  legd_guide <- guide_legend(override.aes = list(
    size   = 3,
    shape  = 21,
    color  = "black",
    stroke = 0.25
  ))
  
  text_theme <- theme(
    axis.title  = element_text(size = 10),
    legend.text = element_text(size = 8),
    axis.text   = element_text(size = 8)
  )
  
  # Reference UMAP
  if (!is.null(on_top)) {
    on_top <- ref_cols[on_top]
    
    ref_cols <- ref_cols[!names(ref_cols) %in% names(on_top)]
    ref_cols <- c(ref_cols, on_top)
  }
  
  ref_umap <- sobj_in %>%
    create_ref_umap(
      feature     = "subtype",
      color_guide = legd_guide,
      plot_cols   = ref_cols,
      pt_mtplyr   = pt_size / 0.1,
      pt_outline  = pt_outline,
      feat_levels = names(ref_cols)
    ) +
    theme(
      legend.position = "left",
      legend.text = element_text(size = 8)
    ) +
    theme(...)
  
  # Create list of feature UMAPs
  feat_umaps <- sobj_in %>%
    create_marker_umaps(
      pt_mtplyr       = pt_size / 0.1,
      add_outline     = pt_outline,
      input_markers   = feats,
      low_col         = low_col,
      high_col_mtplyr = high_col_mtplyr
    ) %>%
    map(~ {
      .x +
        guides(color = guide_colorbar(frame.colour = "black", frame.linewidth = 0.2)) +
        theme(plot.title = element_text(size = 12))
    })
  
  # Return plot list
  if (return_list) {
    return(append(list(ref_umap), feat_umaps))
  }
  
  # Top panel of feature UMAPs
  n_umaps <- (length(feat_umaps) + 2) / n_row
  n_top <- n_umaps - 2
  
  top_umaps <- append(list(ref_umap), feat_umaps[1:n_top]) %>%
    plot_grid(
      plotlist   = .,
      rel_widths = c(1, rep(0.5, n_top)),
      nrow       = 1
    )
  
  # Bottom panel of feature UMAPs
  bot_umaps <- feat_umaps[(n_top + 1):length(feat_umaps)] %>%
    plot_grid(
      plotlist = .,
      nrow     = n_row - 1,
      align    = "h",
      axis     = "tb"
    )
  
  # Final feature UMAP figure
  res <- plot_grid(
    top_umaps, bot_umaps,
    rel_heights = c(1, n_row - 1),
    ncol        = 1,
    align       = "vh",
    axis        = "trbl"
  ) + 
    theme(plot.margin = unit(c(1, 0.2, 1.5, 0.2), "cm"))
  
  res
}

# Create figure 4 panels
create_fig4 <- function(sobj_in, feat_cols, ref_cols, ova_cols, pt_size = 0.00001, pt_outline = 0.4, median_pt = 1,
                        low_col = c("white", "white"), high_col_mtplyr = 1, on_top = NULL,  gmm_column = "GMM_grp",
                        gmm_grps = c("ova low", "ova high")) {
  
  # Theme elements
  feats <- names(feat_cols)
  
  text_theme <- theme(
    axis.title  = element_text(size = 10),
    legend.text = element_text(size = 8),
    axis.text   = element_text(size = 8)
  )
  
  # Set GMM groups in Seurat object
  sobj_in@meta.data <- sobj_in@meta.data %>%
    rownames_to_column("cell_id") %>%
    mutate(!!sym(gmm_column) := ifelse(
      !(!!sym(gmm_column)) %in% gmm_grps,
      "Other",
      !!sym(gmm_column)
    )) %>%
    column_to_rownames("cell_id")
  
  # GMM summary plots
  gmm_panels <- create_gmm_panels(
    sobj_in     = sobj_in,
    gmm_column  = gmm_column,
    ova_cols    = ova_cols,
    bar_cols    = ref_cols,
    pt_size     = pt_size,
    pt_outline  = pt_outline,
    show_bars   = T,
    plot_labs   = letters[1:3],
    rel_heights = c(1, 0.65, 0.7), 
    nrow        = 3,
    align       = "v",
    axis        = "l"
  )
  
  # Feature UMAPs
  feat_umaps <- create_feat_umap_panels(
    sobj_in           = sobj_in,
    ref_cols          = ref_cols,
    feats             = feat_cols,
    pt_size           = pt_size,
    pt_outline        = pt_outline,
    low_col           = low_col,
    high_col_mtplyr   = high_col_mtplyr,
    on_top            = on_top,
    return_list       = T,
    legend.margin     = margin(0.2, 0, 0.2, 1, "cm"),
    legend.key.height = unit(0.4, "cm")
  )
  
  umap_legd <- get_legend(feat_umaps[[1]]) %>%
    plot_grid(
      ., NULL,
      rel_heights = c(0.4, 1),
      ncol = 1
    )
  
  feat_umaps[[1]] <- feat_umaps[[1]] +
    theme(legend.position = "none")
  
  feat_umaps <- feat_umaps %>%
    plot_grid(
      plotlist = .,
      ncol     = 2,
      align    = "vh",
      axis     = "trbl"
    )
  
  feat_umaps <- feat_umaps %>%
    plot_grid(
      umap_legd, .,
      rel_widths = c(0.1, 1),
      nrow = 1
    )
  
  # Boxplot data
  # What to arrange boxplots by median and 3rd quartile
  box_data <- sobj_in %>%
    FetchData(c(feats, "subtype", gmm_column)) %>%
    as_tibble(rownames = "cell_id") %>%
    pivot_longer(cols = c(-cell_id, -subtype, -!!sym(gmm_column))) %>%
    mutate(type_name = str_c(subtype, "_", name)) %>%
    group_by(type_name) %>%
    mutate(
      up_qt = boxplot.stats(value)$stats[4],
      med = median(value)
    ) %>%
    ungroup() %>%
    arrange(desc(med), desc(up_qt)) %>%
    mutate(
      name = fct_relevel(name, feats),
      type_name = fct_inorder(type_name)
    )
  
  # Create boxplots
  feat_boxes <- box_data %>%
    ggplot(aes(type_name, value, fill = subtype)) +
    facet_wrap(~ name, ncol = 1, scales = "free_x") +
    scale_fill_manual(values = ref_cols) +
    scale_color_manual(values = ref_cols) +
    labs(y = "Counts") +
    theme_minimal_hgrid() +
    text_theme +
    theme(
      strip.text         = element_text(size = 12),
      legend.position    = "bottom",
      legend.title       = element_blank(),
      legend.key.height  = unit(0.35, "cm"),
      axis.title.x       = element_blank(),
      axis.text.x        = element_blank(),
      axis.line.x        = element_blank(),
      axis.ticks.x       = element_blank(),
      axis.line.y        = element_line(size = 0.5, color = "grey85"),
      panel.grid.major.y = element_blank(),
      plot.margin        = unit(c(0.1, 2, 0.1, 1.5), "cm")
    )
  
  feat_boxes <- feat_boxes +
    stat_summary(geom = "point", shape = 22, fun = median, size = 0) +
    stat_summary(geom = "point", shape = 22, fun = median, size = median_pt + 1, color = "black") +
    geom_boxplot(
      size           = 0.3,
      width          = 0.6,
      fatten         = 0,
      outlier.colour = "grey85",
      outlier.alpha  = 1,
      outlier.size   = 0.1,
      show.legend    = F,
      coef           = 0,
    ) +
    stat_summary(
      aes(color = subtype),
      geom        = "point",
      shape       = 22,
      fun         = median,
      size        = median_pt,
      stroke      = 1,
      fill        = "white",
      show.legend = F
    ) +
    guides(fill = guide_legend(ncol = 2, override.aes = list(
      size = 3.5,
      stroke = 0.25
    )))
  
  # Create final figure
  res <- plot_grid(
    gmm_panels, feat_umaps, feat_boxes,
    rel_widths     = c(0.8, 1, 0.6),
    nrow           = 1,
    labels         = c("", "d", "e"),
    label_fontface = "plain",
    label_size     = 18
  )
  
  res
}

# Create figure S10 panels
create_figS10 <- function(sobj_in, feat_cols, ref_cols, ova_cols, pt_size = 0.00001, pt_outline = 0.4, n_row = 2,
                          low_col = c("white", "white"), high_col_mtplyr = 1, on_top = NULL,  gmm_column = "GMM_grp",
                          gmm_grps = c("ova low", "ova high"), rel_h = c(0.55, 1.05), plot_labs, ...) {
  
  # Set GMM groups in Seurat object
  sobj_in@meta.data <- sobj_in@meta.data %>%
    rownames_to_column("cell_id") %>%
    mutate(!!sym(gmm_column) := ifelse(
      !(!!sym(gmm_column)) %in% gmm_grps,
      "Other",
      !!sym(gmm_column)
    )) %>%
    column_to_rownames("cell_id")
  
  # GMM summary plots
  gmm_panels <- create_gmm_panels(
    sobj_in    = sobj_in,
    gmm_column = gmm_column,
    ova_cols   = ova_cols,
    bar_cols   = ref_cols,
    pt_size    = pt_size,
    pt_outline = pt_outline,
    show_bars  = F,
    plot_labs  = c(plot_labs[1:2], "", ""), 
    rel_widths = c(0.86, 1, 0.7, 0.2),
    nrow       = 1,
    align      = "h",
    axis       = "tb"
  )
  
  # Feature UMAPs
  feat_umaps <- create_feat_umap_panels(
    sobj_in         = sobj_in,
    ref_cols        = ref_cols,
    feats           = feat_cols,
    n_row           = n_row,
    pt_size         = pt_size,
    pt_outline      = pt_outline,
    on_top          = on_top,
    low_col         = low_col,
    high_col_mtplyr = high_col_mtplyr,
    legend.margin   = margin(0.2, 0.2, 0.2, 1.5, "cm"),
    # plot.margin     = unit(c(0.2, 1.5, 0.2, 0.2), "cm"),
    ...
  )
  
  res <- plot_grid(
    gmm_panels, feat_umaps,
    rel_heights    = rel_h,
    ncol           = 1,
    labels         = c("", plot_labs[3:4]),
    label_fontface = "plain",
    label_size     = 18
  )
  
  res
}

# Create violin plots for DNA-tag counts
create_count_violins <- function(sobj_in, plot_cols, box_counts, type_column = "subtype",
                                 plot_title = NULL, control_types = c("B cell", "T cell")) {
  
  # Violin plot helper
  create_violins <- function(df_in, filt = NULL, color_id = type_column, plot_cols,
                             gg_title = NULL, plot_levels, as_swarm = F) {
    
    if (!is.null(filt)) {
      df_in <- df_in %>%
        filter(name == filt)
    }
    
    res <- df_in %>%
      mutate(!!sym(color_id) := fct_relevel(!!sym(color_id), plot_levels)) %>%
      ggplot(aes(counts + 1, !!sym(color_id))) +
      facet_wrap(~ name) +
      theme_minimal_vgrid() +
      theme(
        legend.position    = "none",
        plot.margin        = unit(c(0.2, 0.2, 1, 0.2), "cm"),
        plot.title         = element_text(size = 12, face = "plain"),
        strip.text         = element_text(size = 10),
        axis.title.y       = element_blank(),
        axis.title         = element_text(size = 10),
        axis.text          = element_text(size = 8),
        axis.ticks.x       = element_line(size = 0.1),
        panel.grid.major.x = element_line(size = 0.1)
      )
    
    if (as_swarm) {
      res <- res +
        geom_quasirandom(aes(color = !!sym(color_id)), size = 0.5, groupOnX = F) +
        scale_color_manual(values = plot_cols)
      
    } else {
      res <- res +      
        geom_violin(aes(fill = !!sym(color_id)), size = 0.3, draw_quantiles = c(0.25, 0.75), alpha = 0.75) +
        stat_summary(geom = "point", color = "black", fun = median) +
        scale_fill_manual(values = plot_cols)
    }
    
    if (!is.null(gg_title)) {
      res <- res +
        ggtitle(gg_title)
      
    }
    
    res
  }
  
  # Data for violin and beeswarm plots
  box_data <- sobj_in %>%
    FetchData(slot = "counts", vars = c(type_column, box_counts)) %>%
    as_tibble(rownames = "cell_id") %>%
    rename(all_of(box_counts)) %>%
    pivot_longer(cols = c(-cell_id, -!!sym(type_column)), values_to = "counts")
  
  # Plot levels
  box_levels <- box_data %>%
    filter(name == names(box_counts[1])) %>%
    mutate(!!sym(type_column) := fct_reorder(!!sym(type_column), counts, median)) %>%
    pull(type_column) %>%
    levels()
  
  box_levels <- box_levels[!box_levels %in% control_types] %>%
    c(control_types, .)
  
  # Create violin plots
  plot_args <- list(
    filt       = names(box_counts),
    as_swarm   = c(F, T, T),
    gg_title   = list(plot_title, NULL, NULL)
  )
  
  vlns <- plot_args %>%
    pmap(
      create_violins,
      df_in       = box_data,
      plot_cols   = plot_cols,
      plot_levels = box_levels
    )
  
  # Set equal x-axis scales
  vlns <- vlns %>%
    equalize_x(log_tran = T)
  
  # Create final figure
  res <- plot_grid(
    plotlist = vlns,
    nrow       = 1,
    align      = "vh",
    axis       = "trbl"
  )
  
  res
}

# Create correlation heatmaps for cell types
create_ref_heatmap <- function(sobj_in, ref_in, type_column = "subtype", plot_cols = NULL, rename_ref = NULL,
                               exclude_ref = "^Endothelial cells ", plot_title = NULL, return_mat = F) {
  
  # Get clustifyr matrix
  cor_mat <- sobj_in %>%
    clustify(
      cluster_col = type_column,
      ref_mat     = ref_in,
      seurat_out  = F
    )
  
  if (return_mat) {
    return(cor_mat)
  }
  
  # Heatmaps
  cor_df <- cor_mat %>%
    as_tibble(rownames = "Cell type") %>%
    rename(all_of(rename_ref)) %>%
    pivot_longer(cols = -`Cell type`, names_to = "Reference type", values_to = "r") %>%
    mutate(
      `Reference type` = fct_reorder(`Reference type`, r, mean, .desc = T),
      `Cell type` = fct_reorder(`Cell type`, r, mean)
    )
  
  # Exclude select cell types from reference
  if (!is.null(exclude_ref)) {
    cor_df <- cor_df %>%
      filter(!grepl(exclude_ref, `Reference type`))
  }
  
  # Create heatmap
  res <- cor_df %>%
    ggplot(aes(`Reference type`, `Cell type`, fill = r)) +
    geom_tile(color = "white", size = 1) +
    scale_fill_gradient(low = "white", high = plot_cols) +
    guides(fill = guide_colorbar(barwidth = unit(0.2, "cm"), frame.colour = "black", frame.linewidth = 0.1)) +
    theme_info +
    theme(
      plot.title   = element_text(size = 12, face = "plain"),
      strip.text   = element_text(size = 10),
      legend.title = element_text(size = 10),
      legend.text  = element_text(size = 8),
      axis.title   = element_text(size = 10),
      axis.text.x  = element_text(angle = 45, hjust = 1),
      axis.text    = element_text(size = 8),
      axis.ticks   = element_blank(),
      axis.line    = element_blank()
    )
  
  
  if (!is.null(plot_title)) {
    res <- res +
     ggtitle(plot_title)
  }
  
  res
}

# Plot correlation
plot_corr <- function(sobj_in, x, y, feat, data_slot, cols_in, plot_title,
                      add_y_pseudo = 1, ...) {
  
  res <- sobj_in %>%
    FetchData(c(x, y, feat), slot = data_slot) %>%
    mutate(
      !!sym(x) := log10(!!sym(x)),
      !!sym(y) := log10(!!sym(y) + add_y_pseudo)
    ) %>%
    plot_features(
      x         = x,
      y         = y,
      feature   = feat,
      data_slot = "counts",
      plot_cols = cols_in,
      lab_pos   = "strip",
      lab_size  = 5,
      calc_cor  = T,
      lm_line   = T,
      ...
    ) +
    ggtitle(plot_title) +
    guides(color = guide_legend(override.aes = list(size = 3.5))) +
    theme_info +
    theme(legend.title = element_blank())
  
  res
}

# Calculate pairwise p-values for Seurat objects
calc_p_vals <- function(sobj_in, sample_name, data_column, type_column, log_tran = F) {
  
  # Helper to calc p-values for give cell types
  calc_p <- function(x, y, df_in, data_column, type_column) {
    
    x_dat <- df_in %>%
      filter(!!sym(type_column) == x) %>%
      pull(data_column)
    
    y_dat <- df_in %>%
      filter(!!sym(type_column) == y) %>%
      pull(data_column)
    
    res <- wilcox.test(
      x        = x_dat, 
      y        = y_dat, 
      conf.int = T
    ) %>%
      tidy() %>%
      mutate(
        group1 = x,
        group2 = y
      )
    
    res
  }
  
  # Pull data from gg object
  p_data <- sobj_in@meta.data %>%
    as_tibble(rownames = "cell_id") %>%
    mutate("cell_type" = !!sym(type_column))
  
  # Log transform
  if (log_tran) {
    p_data <- p_data %>%
      mutate(!!sym(data_column) := log10(!!sym(data_column)))
  }
  
  # Calculate additional stats for table
  p_stats <- p_data %>%
    group_by(cell_type) %>%
    summarize(
      n_cells = n_distinct(cell_id),
      med = median(!!sym(data_column))
    ) %>%
    ungroup() %>%
    mutate(frac_cells = n_cells / sum(n_cells))
  
  # Create data.frame with all combinations of cell types for comparison
  # Flip group columns for B and T cells so group1 includes all cell types
  c_types <- p_data %>%
    pull(type_column) %>%
    unique()
  
  c_comps <- combinations(
    n = length(c_types),
    r = 2,
    v = c_types
  ) %>%
    as_tibble(.name_repair = "unique") %>%
    mutate(
      ...1 = if_else(...1 == "B cell" & ...2 == "T cell", "T cell", ...1),
      ...2 = if_else(...1 == "T cell" & ...2 == "T cell", "B cell", ...2)
    )

  # Calculate pairwise wilcox test for all combinations of cell types
  res <- map2(c_comps$...1, c_comps$...2, ~ {
    calc_p(
      x           = .x,
      y           = .y,
      df_in       = p_data,
      data_column = data_column,
      type_column = type_column
    )
  }) %>%
    bind_rows()
  
  # Apply multiple testing correction
  res <- res %>%
    mutate(p_adj = p.adjust(p.value, method = "bonf"))
  
  # Add medians to data.frame
  res <- p_stats %>%
    rename(med_1 = med) %>%
    right_join(res, by = c(cell_type = "group1"))
  
  res <- p_stats %>%
    select(cell_type, med_2 = med) %>%
    right_join(res, by = c("cell_type" = "group2"))
  
  # Format final table
  res <- res %>%
    mutate(Sample = sample_name) %>%
    select(
      Sample,
      `Cell type 1`      = cell_type.y,
      `n cells 1`        = n_cells,
      `frac cells 1`     = frac_cells,
      `Median rel ova 1` = med_1,
      `Cell type 2`      = cell_type,
      `Median rel ova 2` = med_2,
      method,
      alternative,
      statistic,
      estimate,
      conf.low,
      conf.high,
      p.value,
      p_adj
    ) %>%
    arrange(`Cell type 1`, p_adj)
  
  res
}

# Helper to find subtype ova markers
find_ova_markers <- function(sobj_in, prefix = "", gmm_column, so_title, p_max = 0.05, FC_min = 0.25,
                             auc_min = 0.5, pct_in_min = 50, pct_out_max = 100, uniq = F, filt_ova_high = T) {
  
  grps <- c("ova low", "ova high") %>%
    set_names(., .)
  
  if (prefix != "") {
    grps <- grps %>%
      map_chr(~ str_c(prefix, "-", .x))
  }
  
  res <- find_markers(
    sobj_in      = sobj_in,
    group_column = gmm_column,
    groups_use   = grps,
    p_max        = p_max,
    FC_min       = FC_min,
    auc_min      = auc_min,
    pct_in_min   = pct_in_min,
    pct_out_max  = pct_out_max,
    uniq         = uniq
  )
  
  if (filt_ova_high) {
    res <- res %>%
      filter(group == grps[["ova high"]])
  }
  
  res <- res %>%      
    arrange(desc(logFC)) %>%
    mutate(Sample = so_title) %>%
    relocate(Sample, before = feature)
  
  res
}

# Helper to write excel files
write_xlsx <- function(df_in, file_in, ...) {
  
  if (file.exists(file_in)) {
    stop("File exists: ", file_in)
  }
  
  df_in %>%
    as.data.frame() %>%
    write.xlsx(
      file      = file_in, 
      row.names = F,
      ...
    )
}

# Create Cell Browser files
create_browser <- function(sobj_in, markers, browser_dir, browser_name, cols = NULL, ...) {

  # Set file paths
  brow_path   <- file.path(browser_dir, browser_name)  
  marker_path <- file.path(brow_path, "markers.tsv")
  
  dir.create(brow_path, recursive = T)
  
  # Create tsv file for marker genes
  markers %>%
    select(
      cluster = group,
      gene    = feature,
      score   = logFC
    ) %>%
    write_tsv(marker_path)
  
  # Change adt_ovalbumin column name so raw counts are pulled
  sobj_in@meta.data <- sobj_in@meta.data %>%
    rename(ova_counts = adt_ovalbumin)
  
  # Create Cell Browser files for gene expression data
  sobj_in %>%
    ExportToCellbrowser(
      markers.file     = marker_path,
      dir              = brow_path,
      reductions       = "umap",
      dataset.name     = browser_name,
      GMM_grp          = "Cluster",
      subtype          = "Cell type",
      nCount_RNA       = "RNA UMI count",
      nFeature_RNA     = "Gene count",
      Percent_mito     = "Percent mito",
      ova_counts       = "ova counts",
      ova_fc           = "Relative ova signal",
      ...
    )
  
  # Save color.tsv
  if (!is.null(cols)) {
    cols %>%
      write_tsv(file.path(brow_path, "colors.tsv"), col_names = F)
  }
}

```

```{r "Setup", echo = F}

# Seurat objects used for subtype annotations
sobjs_sub <- read_rds(here(params$rds_dir, "sobjs_subsets.rds"))

# Order of UMAP layers
subtype_order <- list(
  d2_DC   = c("CCR7hi Mig cDC2", "CCR7hi Mig cDC1", "B cell", "T cell"),
  d14_DC  = c("CCR7hi Mig cDC2", "CCR7hi Mig cDC1", "cDC2 Tbet-", "cDC2 Tbet+", "B cell", "T cell", "NK"),
  d2_LEC  = NULL,
  d14_LEC = c("Ptx3_LEC", "fLEC", "cLEC", "Collecting", "BEC", "B cell", "T cell"),
  d2_FRC  = NULL,
  d14_FRC = NULL
)

# Write excel files
write_tables <- !is.null(params$xlsx_dir)

```

# Methods

Fastq files from the gene expression and antigen tracking libraries were processed using the feature barcode version of the cellranger count pipeline (v3.1.0). Reads were aligned to the mm10 and Vaccinia virus (NC_006998) reference genomes. Analysis of gene expression and antigen tracking data was performed using the Seurat R package (v3.2). Antigen tracking and gene expression data were combined into the same Seurat object for each sample (CD45-/day2, CD45+/day2, CD45-/day14, CD45+/day14). Cells were filtered based on the number of detected genes (>250 and <5000) and the percent of mitochondrial reads (<15%). Gene expression counts were log nor-malized (NormalizeData), and relative ova signal was calculated by dividing ova-psDNA counts by the median ova-psDNA counts for all T and B cells present in the sample. Gene expression data were scaled and centered (ScaleData). 2000 variable features (FindVariableFeatures) were used for PCA (RunPCA) and the first 40 principal components were used to find clusters (FindNeighbors, FindClusters) and calculate uniform manifold approximation and projection (UMAP) (RunUMAP). Cell types were annotated using the R package clustifyr (https://rnabi-oco.github.io/clustifyr) along with reference bulk RNA-seq data from ImmGen (available for download through the clustifyrdata R package, https://rnabioco.github.io/clustifyrdata). The samples were then divided into separate objects for DCs, LECs, and FRCs. Cell subsets were annotated using clustifyr with reference bulk RNA-seq data for DCs, LECs, and FRCs.

Identification of ova-low and ova-high populations was accomplished using a gaussian mixture model implemented with the R package mixtools (https://cran.r-project.org/web/packages/mix-tools/index.html). All LECs were used when identifying ova-low and ova-high cells (Fig. 4). For DCs (Fig. S10), ova-low and -high populations were identified independently for each DC cell type. For ova-low and ova-high populations, differentially expressed genes were identified using the R package presto (wilcoxauc, https://github.com/immunogenomics/presto). Differentially expressed genes were filtered to include those with an adjusted p-value <0.05, log fold change >0.25, AUC >0.5, and with at least 50% of ova-high cells expressing the gene. A reproducible analysis pipeline is available at https://github.com/rnabioco/antigen-tracking.

\newgeometry{top=1in,bottom=1in,left=1.5in,right=1.5in}

# Figure 3

Cell types associated with high antigen counts. (a, d, g, j) UMAP projections are shown for DCs, LECs, and FRCs. (b, e, h, k) Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c, f, i, l) Antigen counts are displayed on UMAP projections for each cell type.

```{r "Fig 3", fig.width = 13, fig.height = 18}

# Parameter lists
fig3_names <- c("d2_DC", "d14_DC", "d14_LEC", "d14_FRC")
fig3_sobjs <- sobjs[fig3_names]

fig3_params <- list(
  sobj_in    = fig3_sobjs, 
  plot_title = so_titles[names(fig3_sobjs)],
  cols_in    = so_cols[names(fig3_sobjs)],
  on_top     = subtype_order[names(fig3_sobjs)]
)

# Create panel plots
gg_list <- fig3_params %>%
  pmap(
    create_fig3,
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    pt_size_2    = 0.1,
    pt_outline_2 = 0.4
  ) %>%
  flatten()

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2

gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equalize_x(log_tran = T)

# Create final figure
fig3 <- gg_list %>%
  wrap_plots(ncol = 3, widths = c(1, 0.75, 1)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(-0.08, 1)
  )

fig3

```

\restoregeometry

# Figure 4

Genes associated with high antigen counts for day 14 LECs. (a) LECs containing low and high antigen counts were identified using a gaussian mixture model. A UMAP projection is shown for ova-low and ova-high cells. T cells, B cells, and epithelial cells are shown in white (Other). (b) The distribution of ova antigen counts is shown for ova-low and ova-high cells. Dotted lines indicate the mean counts for each population. (c) The fraction of cells belonging to each LEC cell type is shown for ova-low and ova-high populations. (d) UMAP projections show the expression of top ova-high markers. (e) The expression of ova-high markers was plotted for each cell type.

```{r "Fig 4", fig.width = 12, fig.height = 8}

# Feature colors
feat_cols <- c(
  "#D7301F", "#0072B2",
  "#009E73", "#4C3250",
  "#E69F00", "#875C04"
)

# Genes of interest
genes <- c("Prox1", "Cavin1", "Cavin2", "Stab1", "Stab2") %>%
  set_names(feat_cols[seq_along(.)], .)

# Color scales for UMAPs
low_cols <- c("white", "white")

mtplyr <- rep(1, length(genes))
mtplyr[5] <- 2

# Seurat object for figure
so_name <- "d14_LEC"

fig_4 <- create_fig4(
  sobj_in         = sobjs[[so_name]],
  feat_cols       = genes,
  ref_cols        = so_cols[[so_name]],
  ova_cols        = ova_cols,
  low_col         = low_cols,
  high_col_mtplyr = mtplyr,
  on_top          = subtype_order[[so_name]],
  gmm_column      = "GMM_grp",
  gmm_grps        = c("ova low", "ova high")
)

fig_4

```

\newgeometry{top=1in,bottom=1in,left=1.5in,right=1.5in}

# Figure S5

Comparison of ova-psDNA, psDNA, and pDNA signal for each cell type. Counts are shown for ova-psDNA, psDNA, and pDNA for DCs (a, b), LECs (c), and FRCs (d).  

```{r "Fig S5", fig.width = 11, fig.height = 15}

box_counts <- c(
  "ovalbumin" = "adt_ovalbumin",
  "psDNA"     = "adt_protected-DNA",
  "pDNA"      = "adt_unprotected-DNA"
)

fig_S5 <- sobjs[fig3_names] %>%
  imap(~ {
    create_count_violins(
      sobj_in     = .x,
      plot_cols   = so_cols[[.y]],
      box_counts  = box_counts,
      type_column = "subtype",
      plot_title  = so_titles[[.y]]
    )
  }) %>%
  plot_grid(
    plotlist       = .,
    ncol           = 1,
    labels         = letters,
    label_fontface = "plain",
    label_size     = 18,
    align          = "vh",
    axis           = "trbl"
  )

fig_S5


```

\newgeometry{top=1in,bottom=1in,left=2in,right=2in}

# Figure S6

Antigen counts were compared with total mRNA counts for each cell for DCs (a, b), LECs (c, d), and FRCs (e, f).

```{r "Fig S6", fig.width = 20, fig.height = 34}

# Parameter lists
so_names <- names(sobjs)

corr_params <- list(
  sobj_in    = sobjs[so_names],
  plot_title = so_titles[so_names], 
  cols_in    = so_cols[so_names]
)

# Padding to add to figure panels
pad <- c(9.5, 2, 9.8, 2, 9.8, 2)

# Create scatter plots
cor_plots <- corr_params %>%
  pmap(
    plot_corr,
    x         = c("RNA counts (log10)" = "nCount_RNA"),
    y         = c("ova counts + 1 (log10)" = "adt_ovalbumin"),
    feat      = "subtype",
    split_id  = "subtype",
    data_slot = "counts",
    pt_size   = 1,
    scales    = "free",
    ncol      = 3
  ) %>%
  map2(pad, ~ {
    .x + theme(
      legend.position = "none",
      strip.text      = element_text(size = 14),
      plot.margin     = unit(c(0.2, 0.2, .y, 0.2), "cm")
    )
  })

# Create final figure
figS6 <- plot_grid(
  plotlist       = cor_plots,
  rel_heights    = c(1, 0.8, 0.8),
  labels         = letters,
  label_size     = 32,
  label_fontface = "plain",
  align          = "v",
  ncol           = 2
)

figS6

```

\restoregeometry

# Figure S7

LEC cell types associated wih high antigen counts for the day 2 timepoint. (a) A UMAP projection is shown for LEC cell types, epithelial cells, B cells, and T cells identified for the day 2 timepoint. (b) Relative ova signal is shown for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c) Antigen counts are displayed on the UMAP projection shown in (a). (d, e) Correlation coefficients are shown comparing each identified LEC cell type with the reference cell types from Xiang et al. 

```{r "Fig S7", fig.width = 13, fig.height = 9}

# Create UMAPs and violin plots for d2_LEC
so_name <- "d2_LEC"

gg_list <- sobjs[[so_name]] %>%
  create_fig3(
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    cols_in      = so_cols[[so_name]],
    plot_title   = so_titles[[so_name]],
    on_top       = subtype_order[[so_name]],
    pt_size      = 0.5,
    pt_outline   = 0.8,
    pt_size_2    = 0.5,
    pt_outline_2 = 0.8
  )

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2
gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equalize_x(log_tran = T)

# Create correlation heatmaps for LEC annotations
LEC_names <- c("d2_LEC", "d14_LEC")
heat_cols <- c(get_cols()[2], get_cols()[3])

heat_maps <- map2(LEC_names, heat_cols, ~ {
  create_ref_heatmap(
    sobj_in    = sobjs_sub[[.x]],
    ref_in     = subtype_refs[[so_types[.x]]],
    plot_cols  = .y,
    plot_title = so_titles[[.x]],
    rename_ref = c("BEC (ImmGen)" = "BEC")
  )
})

# Create final figure
fig_S7 <- (gg_list[[1]] | gg_list[[2]] | gg_list[[3]]) /
  (heat_maps[[1]] | heat_maps[[2]]) +
  plot_layout(heights = c(1, 0.6)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(0, 1)
  )

fig_S7

```

\newpage

# Figure S9

FRC cell types associated wih high antigen counts for the day 2 timepoint. (a) A UMAP projection is shown for FRC cell types, B cells, and T cells identified for the day 2 timepoint. (b) Relative ova signal is shown for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c) Antigen counts are displayed on the UMAP projection shown in (a).

```{r "Fig S9", fig.width = 13, fig.height = 5}

# Create UMAPs and violine plots for d2_FRC
so_name <- "d2_FRC"

gg_list <- sobjs[[so_name]] %>%
  create_fig3(
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    cols_in      = so_cols[[so_name]],
    plot_title   = so_titles[[so_name]],
    on_top       = subtype_order[[so_name]],
    pt_size      = 1.5,
    pt_outline   = 2,
    pt_size_2    = 1.5,
    pt_outline_2 = 2
  )

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2
gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equalize_x(log_tran = T)

# Create final figure
fig_S9 <- gg_list %>%
  wrap_plots(ncol = 3, widths = c(1, 0.75, 1)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(-0.08, 1)
  )

fig_S9

```

\newgeometry{top=0.5in,bottom=0.5in,left=2.3in,right=2.3in}

# Figure S10

Genes associated with high antigen counts for day 2 and day 14 cDC2 Tbet- cells. (a, d) Day 2 (a) and day 14 (d) cDC2 Tbet- cells containing low and high antigen counts were identified using a gaussian mixture model. A UMAP projection is shown for ova-low and ova-high cells. Cell types not included in the comparison are shown in white (Other). (b, e) The distribution of ova antigen counts is shown for ova-low and ova-high cDC2 Tbet- cells. Dotted lines indicate the mean counts for each population. (c, f) UMAP projections show the expression of top markers associated with ova-high cDC2 Tbet- cells.

```{r "Fig S10", fig.width = 10, fig.height = 20}

# Seurat objects, genes, cell types
so_names <- c("d2_DC", "d14_DC")
type <- "cDC2 Tbet-"

genes <- list(
  d2_DC = c(
    "Ccl2", "Cxcl2", "Mif", "Msr1",
    "Pkm", "Lgals3", "Lgmn", "Il1rn",
    "Dusp1", "Maf", "Fpr2", "Ctsb",
    "Clec4a1"
  ),
  d14_DC = c(
    "Fcgr4", "Cd72", "Prkcb", "Acp5",
    "Itgam", "Pecam1", "Siglece", "Ccnd2",
    "Cd300e", "C1qb", "Tcf7l2", "Pltp",
    "Fcrl5"
  )
) %>%
  map(set_names, x = get_cols()[c(1:3, 5:14)])

# OVA colors and groups
fig_cols <- rev(ova_cols)

names(fig_cols) <- names(fig_cols) %>%
  str_c(if_else(. != "Other", str_c(type, "-"), ""), .)
  
gmm_grps <- names(fig_cols)[names(fig_cols) != "Other"]

# Figure parameters
mtplyr_1 <- rep(1, 13)
mtplyr_1[2] <- 2

plot_args <- list(
  sobj_in         = sobjs[so_names],
  feat_cols       = genes[so_names],
  ref_cols        = so_cols[so_names],
  on_top          = subtype_order[so_names],
  plot_labs       = list(letters[1:3], letters[4:6]),
  low_col         = list(c("white", "white"), "white"),
  high_col_mtplyr = list(mtplyr_1, 1)
)

# Create figures
fig_S10 <- plot_args %>%
  pmap(
    create_figS10,
    ova_cols    = fig_cols,
    gmm_column  = "type_GMM_grp_2",
    gmm_grps    = gmm_grps,
    n_row       = 3,
    rel_h       = c(0.55, 1.4),
    plot.margin = unit(c(0.2, 0, 0.7, 0), "cm")
  ) %>%
  plot_grid(
    plotlist = .,
    ncol     = 1,
    align    = "vh", 
    axis     = "trbl"
  )

fig_S10

```

# Table S2

Comparison of relative ova signal for cell types shown in Fig 3, S7b, and S9b. A two-sided Wilcoxon rank sum test was used to compare relative ova signal for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. The Bonferroni method was used to correct for multiple comparisons. The number of cells in each group (n cells), fraction of the total cells for the sample (frac cells), median relative ova signal, test statistic (statistic), estimation of the median difference (estimate), and confidence interval (conf.low, conf.high) are included.

```{r "Table S2", eval = write_tables}

# Create table of ova counts stats
ova_stats <- sobjs %>%
  imap(~ {
    calc_p_vals(
      sobj_in     = .x,
      sample_name = so_titles[.y],
      data_column = "ova_fc",
      type_column = "subtype"
    )
  }) %>%
  bind_rows()

# Write excel file
xlsx_file <- here(params$xlsx_dir, "table_S2.xlsx")

ova_stats %>%
  write_xlsx(xlsx_file)

```

# Table S3

Genes associated with ova-high cells for DCs, FRCs, and LECs. Ova-high cells were independently identified for DCs, FRCs, and LECs using a gaussian mixture model implemented with the R package mixtools. Differentially expressed genes were identified using a Wilcoxon rank sum test performed using the R package presto (wilcoxauc). The Benjamini-Hochberg method was used to correct for multiple comparisons. Genes were filtered to only include those with an adjusted p-value <0.05, log fold change >0.25, auc >0.5, and with at least 50% of ova-high cells expressing the gene. The average expression, log fold change, test statistic (statistic), area under the receiver operator curve (auc), percentage of ova-high cells that express the gene (pct_in), and percentage of ova-low cells that express the gene (pct_out) are included.

```{r "Table S3", eval = write_tables}

# Ova markers
ova_marks <- sobjs %>%
  imap(~ {
    find_ova_markers(
      sobj_in       = .x,
      prefix        = "",
      gmm_column    = "GMM_grp",
      so_title      = so_titles[.y],
      filt_ova_high = F
    )
  }) %>%
  bind_rows()

# Write excel file
xlsx_file <- here(params$xlsx_dir, "table_S3.xlsx")

ova_marks %>%
  filter(group == "ova high") %>%
  write_xlsx(xlsx_file)

```

# Table S4

Genes associated with ova-high cells for different DC, FRC, and LEC cell types. Ova-high cells were identified independently for each cell type shown in Fig 3 using a gaussian mixture model implemented with the R package mixtools. Differentially expressed genes were identified using a Wilcoxon rank sum test performed using the R package presto (wilcoxauc). The Benjamini-Hochberg method was used to correct for multiple comparisons. Genes were filtered to only include those with an adjusted p-value <0.05, log fold change >0.25, auc >0.5, and with at least 50% of ova-high cells expressing the gene. The average expression, log fold change, test statistic (statistic), area under the receiver operator curve (auc), percentage of ova-high cells that express the gene (pct_in), and percentage of ova-low cells that express the gene (pct_out) are included.

```{r "Table S4", eval = write_tables}

# Subtype ova markers
exclude_types <- c("T cell", "B cell", "Epithelial", "NK")

sub_ova_marks <- sobjs %>%
  imap(~ {
    so_in <- .x
    so_title <- so_titles[.y]
    
    cell_types <- so_in@meta.data$subtype %>%
      unique() %>%
      .[!. %in% exclude_types]
    
    res <- cell_types %>%
      map(~ {
        find_ova_markers(
          sobj_in    = so_in,
          prefix     = .x,
          gmm_column = "type_GMM_grp_2",
          so_title   = so_title
        )
      }) %>%
      bind_rows()
    
    res
  }) %>%
  bind_rows()

# Write excel file
xlsx_file <- here(params$xlsx_dir, "table_S4.xlsx")

sub_ova_marks %>%
  write_xlsx(xlsx_file)

```

```{r "Cellbrowser 1", eval = F}

# Save color tsv
browser_cols <- append(so_cols, ova_cols) %>%
  flatten() %>%
  tibble(
    metadataValue = names(.),
    colorCode = .
  ) %>%
  unnest(cols = colorCode) %>%
  unique() %>%
  mutate(colorCode = if_else(
    metadataValue == "Other", 
    "#f0f0f0", 
    colorCode
  ))

# Create cellbrowser directories
sobjs %>%
  iwalk(~ {
    marks_in  <- ova_marks %>%
      filter(Sample == so_titles[[.y]])
    
    create_browser(
      sobj_in      = .x,
      markers      = marks_in,
      browser_dir  = params$browser_dir,
      browser_name = .y,
      cols         = browser_cols,
      skip.expr.matrix = T
    )
  })

```

```{bash "Cellbrowser 2", eval = F}

html_dir=../cellbrowser

mkdir -p $html_dir/html

browsers=(
  d2_DC  d14_DC
  d2_LEC d14_LEC
  d2_FRC d14_FRC
)

args=""

for dir in ${browsers[@]}
do
  echo -e "\ncolors=\"colors.tsv\"" \
    >> $html_dir/$dir/cellbrowser.conf

  args=$(echo "$args" "-i $html_dir/$dir/cellbrowser.conf")
done

cbBuild $args -o $html_dir/html

```









