
```{r "{{.x}} Setup", echo = F}

# Setup
rds_names        <- str_split("{{.x}}", ",")[[1]]
rds_files        <- str_c(params$rds_dir, "/", rds_names, "_so.rds")
names(rds_files) <- rds_names

ext_names        <- str_split("{{.y}}", ",")[[1]]
ext_files        <- str_c(params$ext_dir, "/", ext_names)
names(ext_files) <- ext_names

sect_title <- c(rds_names, str_trunc(ext_names, 20))

# Section descriptions
marker_desc <- "Gene expression signal is shown for the top marker genes (top, middle)."
GO_desc <- "GO terms were identified for marker genes, the top terms are labeled, the size of each point indicates the number of overlapping genes (bottom)."

```

```{r "{{.x}} Load data"}

# Load Seurat objects
sobjs <- rds_files %>%
  imap(~ {
    tm        <- str_split(.y, pattern = "_", )[[1]][1]
    type_colm <- str_split(.y, pattern = "_")[[1]][2]
    type_colm <- str_c("cell_type", type_colm)
    res       <- read_rds(.x)
    
    res@meta.data <- res@meta.data %>%
      rownames_to_column("cell_id") %>%
      mutate(
        orig.ident = tm,
        type = !!sym(type_colm),
        type = str_replace(type, " ", "_"),
        type = str_c(tm, "_", type)
      ) %>%
      column_to_rownames("cell_id")
    
    res
  })

# Create Clustifyr reference
subtype_ref <- sobjs %>%
  merge_sobj() %>%
  seurat_ref("type")

# Load external data
sobj_ext <- ext_files %>%
  read.csv() %>%
  column_to_rownames("X") %>%
  CreateSeuratObject()

# ext_meta <- read_tsv("../ext_data/GSE137710_mouse_spleen_cell_metadata_4464x9.tsv.gz") %>%
#   group_by(cell_ID) %>%
#   filter(n() == 1) %>%
#   column_to_rownames("cell_ID")
# 
# sobj_ext <- ext_files %>%
#   read.table() %>%
#   t() %>%
#   CreateSeuratObject(meta.data = ext_meta)

sobj_ext[["percent_mt"]] <- sobj_ext %>%
  PercentageFeatureSet(pattern = "^mt-")

sobj_ext <- sobj_ext %>%
  subset(
    nFeature_RNA > 200 &
    nFeature_RNA < 5000 &
    percent_mt < 15
  )

sobj_ext@meta.data <- sobj_ext@meta.data %>%
  rownames_to_column("cell_id") %>%
  mutate(orig.ident = "ext_data") %>%
  column_to_rownames("cell_id")

# Normalize counts
sobj_ext <- sobj_ext %>%
  NormalizeData(normalization.method = "LogNormalize") %>%
  ScaleData(assay = "RNA") %>%
  FindVariableFeatures(asssay = "RNA")

# Cluster
sobj_ext <- sobj_ext %>%
  cluster_RNA(
    assay      = "RNA",
    resolution = 0.6,
    dims       = 1:40
  ) %>%
  AddMetaData(
    metadata = Embeddings(., reduction = "umap"),
    col.name = c("UMAP_1", "UMAP_2")
  )

# Clustify using our data as reference
sobj_ext <- sobj_ext %>%
  clustify(
    cluster_col = "seurat_clusters",
    ref_mat = subtype_ref
  )

# Create list of all sobjs
all_sobjs <- c(sobjs, ext_data = sobj_ext)

# Merge objects and run PCA, UMAP, clustering
sobj_merge <- all_sobjs %>%
  merge_sobj(sample_order = names(sobjs)) %>%
  cluster_RNA(
    assay      = "RNA",
    resolution = 1,
    dims       = 1:40
  ) %>%
  AddMetaData(
    metadata = Embeddings(., reduction = "umap"),
    col.name = c("merge_UMAP_1", "merge_UMAP_2")
  )

# Integrate samples
sobj_int <- sobj_merge %>%
  SplitObject(split.by = "orig.ident") %>%
  map(FindVariableFeatures) %>%
  FindIntegrationAnchors(dims = 1:40) %>%
  IntegrateData(dims = 1:40) %>%
  ScaleData(assay = "integrated") %>%
  cluster_RNA(
    assay      = "integrated",
    resolution = 1,
    features   = rownames(.),
    prefix     = "int_"
  ) %>%
  AddMetaData(
    metadata = Embeddings(., reduction = "int_umap"),
    col.name = c("UMAP_1", "UMAP_2")
  )

sobj_int@meta.data <- sobj_int@meta.data %>%
  rownames_to_column("cell_id") %>%
  mutate(orig_type = str_c(orig.ident, "-", type)) %>%
  column_to_rownames("cell_id")

```

## **`r sect_title`**

### Sample Markers

UMAPs showing the sample identity are included using data for before (left) and after (right) sample integration.
```{r "{{.x}} Sample summary", fig.width = 8.5, fig.height = 5.8}

# Sample colors
samples <- unique(sobj_int$orig.ident)
sample_cols <- theme_cols[seq_along(samples)]
names(sample_cols) <- samples

# Merged UMAPs
merge_umap <- sobj_int %>%
  plot_features(
    x = "merge_UMAP_1", 
    y = "merge_UMAP_2", 
    feature = "orig.ident",
    plot_cols = theme_cols
  ) +
  ggtitle("Merged") +
  guides(color = guide_legend(nrow = 4, override.aes = list(size = 3.5))) +
  blank_theme +
  theme(
    legend.position = "top",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 10)
  )

legnd <- get_legend(merge_umap)

merge_umap <- merge_umap +
  theme(legend.position = "none")

# Integrated UMAPs
int_umap <- sobj_int %>%
  plot_features(
    x = "UMAP_1", 
    y = "UMAP_2", 
    feature = "orig.ident",
    plot_cols = theme_cols
  ) +
  ggtitle("Integrated") +
  guides(color = guide_legend(override.aes = list(size = 3.5))) +
  blank_theme +
  theme(legend.position = "none")

# Create final figure
umaps <- plot_grid(
  merge_umap, int_umap,
  nrow  = 1,
  align = "vh",
  axis  = "trbl"
)

plot_grid(
  umaps, legnd,
  rel_heights = c(0.8, 0.2),
  ncol = 1
)

```

### {.tabset .tabset-pills}

Marker genes were identified by comparing each sample. `r marker_desc` `r GO_desc`
```{r "{{.x}} Sample markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Create figures
create_marker_panel_v1(
  input_sobj       = sobj_int,
  input_cols       = sample_cols,
  clust_column     = "orig.ident",
  n_boxes          = 24,
  all_violins      = T,
  panel.background = element_rect(color = "#fafafa", fill = "#fafafa"),
  panel.spacing.y  = unit(0, "cm"),
  panel.spacing.x  = unit(0.2, "cm"),
  strip.text       = element_text(size = 10, angle = 90, hjust = 0)
)

```

### Subtype Markers

UMAPs showing cell subtypes are included using data for before (left) and after (right) sample integration.
```{r "{{.x}} Subtype summary", fig.width = 8.5, fig.height = 5.8}

# Subtype colors
cell_types <- sobj_int@meta.data %>%
  arrange(type, orig_type) %>%
  pull(orig_type) %>%
  unique()

type_cols <- theme_cols[seq_along(cell_types)]
names(type_cols) <- cell_types

# Merged UMAPs
merge_umap <- sobj_int %>%
  plot_features(
    x = "merge_UMAP_1", 
    y = "merge_UMAP_2", 
    feature = "type",
    plot_cols = theme_cols
  ) +
  ggtitle("Merged") +
  guides(color = guide_legend(nrow = 4, override.aes = list(size = 3.5))) +
  blank_theme +
  theme(
    legend.position = "top",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 10)
  )

legnd <- get_legend(merge_umap)

merge_umap <- merge_umap +
  theme(legend.position = "none")

# Integrated UMAPs
int_umap <- sobj_int %>%
  plot_features(
    x = "UMAP_1", 
    y = "UMAP_2", 
    feature = "type",
    plot_cols = theme_cols
  ) +
  ggtitle("Integrated") +
  guides(color = guide_legend(override.aes = list(size = 3.5))) +
  blank_theme +
  theme(legend.position = "none")

# Create final figure
umaps <- plot_grid(
  merge_umap, int_umap,
  nrow  = 1,
  align = "vh",
  axis  = "trbl"
)

plot_grid(
  umaps, legnd,
  rel_heights = c(0.8, 0.2),
  ncol = 1
)

```

### {.tabset .tabset-pills}

Marker genes were identified by dividing cells based on subtype and comparing each sample. `r marker_desc` `r GO_desc`
```{r "{{.x}} Subtype markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Find subtype markers
fig_markers <- unique(sobj_int$type) %>%
  map(~ {
    find_group_markers(
      input_grp  = .x,
      input_sobj = sobj_int,
      grp_col    = "type",
      clust_col  = "orig_type"
    )
  }) %>%
  compact() %>%
  bind_rows()

# Create figures
create_marker_panel_v2(
  input_sobj       = sobj_int,
  input_markers    = fig_markers,
  input_cols       = type_cols,
  grp_column       = "type",
  clust_column     = "orig_type", 
  n_boxes          = 24,
  all_violins      = T,
  panel.background = element_rect(color = "#fafafa", fill = "#fafafa"),
  panel.spacing.y  = unit(0, "cm"),
  panel.spacing.x  = unit(0.2, "cm"),
  strip.text       = element_text(size = 10, angle = 90, hjust = 0)
)

```


```{r "EXTRAs", eval = F, echo = F}

# # Find markers
# lec_marks <- lec %>%
#   FindAllMarkers(only.pos = T)
# 
# # Plot known subtype markers
# known_markers <- c(
#   "Ackr4",
#   "Madcam1",
#   "Ccl20",
#   "Cd274",
#   "Lyve1",
#   "Foxc2",
#   "Ccl21a",
#   "Marco",
#   "Ptx3"
# )
# 
# lec %>%
#   FetchData(c("RNA_clusters", known_markers)) %>%
#   as_tibble(rownames = "cell_id") %>%
#   pivot_longer(cols = c(-cell_id, -RNA_clusters), names_to = "gene", values_to = "counts") %>%
#   ggplot(aes(RNA_clusters, counts + 0.1, color = RNA_clusters, fill = RNA_clusters)) +
#   geom_quasirandom(size = 0.5) +
#   facet_wrap(~ gene, scales = "free_y") +
#   scale_y_log10() +
#   scale_fill_manual(values = theme_cols) +
#   scale_color_manual(values = theme_cols) +
#   stat_summary(fun = "median", geom = "point", shape = 22, size = 1, fill = "white", show.legend = F) +
#   guides(color = guide_legend(override.aes = list(size = 3))) +
#   theme_info +
#   theme(legend.title = element_blank())

# Transfer labels
# sobj_labs <- sobj %>%
#   FindTransferAnchors(
#     query = sobj_ext, 
#     dims = 1:30
#   ) %>%
#   TransferData(
#     refdata = sobjs[["our_data"]]$type,
#     dims = 1:30
#   )
# 
# sobj_ext <- sobj_ext %>%
#   AddMetaData(sobj_labs)

# Filter for LECs (Pdpn+/Pecam1+)
# sobj %>%
#   plot_features(
#     x         = "Pdpn",
#     y         = "Pecam1",
#     feature   = "orig.ident",
#     plot_cols = theme_cols,
#     pt_size   = 1
#   ) +
#   theme_info +
#   theme(legend.position = "none")
# 
# lec <- sobj %>%
#   subset(subset = Pecam1 > 0.5 & Pdpn > 0.5)

# Create LEC Clustifyr reference
# Convert human LEC gene names to mouse orthologs
# load("../clustifyr/ref_LEC.rda")
# 
# ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# 
# orthologs <- getBM(
#   mart = ensembl,
#   attributes = c("external_gene_name", "mmusculus_homolog_associated_gene_name")
# )
# 
# mus_genes_LEC <- ref_LEC %>%
#   as_tibble(rownames = "external_gene_name") %>%
#   left_join(orthologs, by = "external_gene_name") %>%
#   dplyr::select(-external_gene_name) %>%
#   filter(
#     !is.na(mmusculus_homolog_associated_gene_name),
#     mmusculus_homolog_associated_gene_name != ""
#   ) %>%
#   dplyr::select(mmusculus_homolog_associated_gene_name, LEC1:"???") %>%
#   group_by(mmusculus_homolog_associated_gene_name) %>%
#   summarise_all(mean)
# 
# mus_ref_LEC <- mus_genes_LEC %>%
#   dplyr::select(-mmusculus_homolog_associated_gene_name) %>%
#   as.matrix()
# 
# rownames(mus_ref_LEC) <- mus_genes_LEC$mmusculus_homolog_associated_gene_name

# Assign cell types
# d2LEC_resLEC_type <- lec %>%
#   FetchData(c("seurat_clusters", "type", "r")) %>%
#   rownames_to_column("cell_id") %>%
#   dplyr::select(-cell_id) %>%
#   unique()
# 
# d2LEC.new.cluster.ids <- c(
#   "LEC3" = "Ceiling LECs", 
#   "CD34+_SC" = "CD34+SC",
#   "BEC"  = "Medullary LECs",
#   "LEC2" = "Medullary LECs",
#   "LEC2" = "Floor LECs"
# )
# 
# x <- lec@meta.data %>%
#   rownames_to_column("cell_id") %>%
#   mutate(type = recode(type, !!!d2LEC.new.cluster.ids))
# 
# LEC_ref <- d2_LEC_so %>%
#   seurat_ref(cluster_col = "cell_typeLEC")
# 
# x@meta.data %>%
#   rownames_to_column("cell_id") %>%
#   mutate(type = recode(type, !!!d2LEC.new.cluster.ids)) %>%
#   head()

```
