
```{r "{{.y}} Setup", echo = F}

# Load Seurat object
rds_type <- "{{.x}}"
rds_name <- "{{.y}}"
rds_file <- str_c(rds_name, "_so.rds")

sobj <- read_rds(file.path(params$rds_dir, rds_file))

mtplyr_1 <- if_else(ncol(sobj) < 500, 3, 1)
mtplyr_2 <- if_else(ncol(sobj) < 500, 6, 1)
mtplyr_3 <- if_else(ncol(sobj) < 500, mtplyr_2 * 2.5, 1)

cell_type_id <- str_c("cell_type", rds_type)
cluster_id   <- str_c(cell_type_id, "2")

# Fix column names
if (!cluster_id %in% names(sobj@meta.data)) {
  sobj@meta.data <- sobj@meta.data %>%
    rownames_to_column("cell_id") %>%
    mutate(!!sym(cluster_id) := str_c(RNA_snn_res.1, "-", !!sym(cell_type_id))) %>%
    column_to_rownames("cell_id")
}

if ("ova" %in% rownames(sobj@assays$adt)) {
  sobj <- sobj %>%
    FetchData("adt_ova") %>%
    AddMetaData(
      object = sobj,
      metadata = .,
      col.name = "adt_ovalbumin"
    )
}

sobj@meta.data <- sobj@meta.data %>%
  rownames_to_column("cell_id") %>%
  mutate(
    cluster_id = !!sym(cluster_id),
    cluster_id = str_replace(cluster_id, "-NA$", "-unassigned"),
    cell_type  = str_remove(cluster_id, "^[0-9]+-")
  ) %>%
  column_to_rownames("cell_id")

# Section descriptions
marker_desc <- "Gene expression signal is shown for the top marker genes (top, middle)."
GO_desc <- "GO terms were identified for marker genes, the top terms are labeled, the size of each point indicates the number of overlapping genes (bottom)."

```

## **`r rds_name`**

### Subtype Markers

A UMAP projection colored by subtype is shown on the left and ovalbumin signal for each subtype is shown on the right.
```{r "{{.y}} Subtype OVA", fig.width = 9, fig.height = 5}

# Set subtype order and colors
cell_types <- sobj@meta.data %>%
  pull(cell_type) %>%
  unique()

names(cell_types) <- cell_types

type_cols <- theme_cols
names(type_cols) <- cell_types
type_cols <- type_cols[!is.na(names(type_cols))]

# Subtype UMAP
type_umap <- sobj %>%
  plot_features(
    feature     = "cell_type",
    pt_size     = 0.1 * mtplyr_3,
    pt_outline  = 0.4,
    plot_cols   = type_cols,
    feat_levels = names(type_cols)
  ) +
  guides(color = guide_legend(override.aes = list(size = 3.5), nrow = 3)) +
  blank_theme +
  theme(
    legend.position = "top",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8)
  )

# OVA boxplots
ova_boxes <- sobj %>%
  FetchData(c("adt_ovalbumin", "cell_type", "cluster_id")) %>%
  as_tibble(rownames = "cell_id") %>%
  mutate(
    adt_ovalbumin = adt_ovalbumin + 0.01,
    cell_type = fct_reorder(cell_type, adt_ovalbumin, median)
  ) %>%
  ggplot(aes(cell_type, adt_ovalbumin, color = cell_type)) +
  geom_quasirandom(size = 0.5 * mtplyr_1) +
  scale_color_manual(values = type_cols) +
  scale_y_log10() +
  coord_flip() +
  theme_info +
  theme(
    legend.position = "none",
    axis.title.y    = element_blank(),
    axis.title.x    = element_text(size = 13),
    axis.text       = element_text(size = 8)
  )

# Create final figure
plot_grid(
  type_umap, ova_boxes,
  rel_widths = c(0.46, 0.54),
  nrow  = 1,
  align = "h",
  axis  = "trbl"
)

```

<br>

### {.tabset .tabset-pills}

Marker genes were identified by comparing all subtypes. `r marker_desc` `r GO_desc`
```{r "{{.y}} Subtype markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Create figures
fig_umap <- type_umap +
  guides(color = col_guide) +
  theme(legend.text = element_text(size = 10))

create_marker_panel_v1(
  input_sobj   = sobj,
  input_cols   = type_cols,
  input_umap   = fig_umap,
  clust_column = "cell_type"
)

```

### Cluster Markers

A UMAP projection colored by cell cluster is shown on the left and ovalbumin signal for each cluster is shown on the right.
```{r "{{.y}} Cluster OVA", fig.width = 9, fig.height = 4.2}

# Set cluster order and colors
cell_clusts <- sobj@meta.data %>%
  mutate(
    type_clust = str_extract(cluster_id, "^[0-9]+"),
    type_clust = as.numeric(type_clust)
  ) %>%
  arrange(cell_type, type_clust) %>%
  pull(cluster_id) %>%
  unique()

names(cell_clusts) <- cell_clusts

clust_cols <- theme_cols
names(clust_cols) <- cell_clusts
clust_cols <- clust_cols[!is.na(names(clust_cols))]

# Cluster UMAP
clust_umap <- sobj %>%
  plot_features(
    feature     = "cluster_id",
    pt_size     = 0.1 * mtplyr_3,
    pt_outline  = 0.4,
    plot_cols   = clust_cols,
    feat_levels = names(clust_cols),
  ) +
  guides(color = guide_legend(override.aes = list(size = 3.5), nrow = 10)) +
  blank_theme +
  theme(legend.position = "none")

# OVA boxes
ova_boxes <- sobj %>%
  FetchData(c("adt_ovalbumin", "cell_type", "cluster_id")) %>%
  as_tibble(rownames = "cell_id") %>%
  mutate(
    adt_ovalbumin = adt_ovalbumin + 0.01,
    cluster_id = fct_reorder(cluster_id, adt_ovalbumin, median)
  ) %>%
  ggplot(aes(cluster_id, adt_ovalbumin, color = cluster_id)) +
  geom_quasirandom(size = 0.25 * mtplyr_2) +
  scale_color_manual(values = clust_cols) +
  scale_y_log10() +
  coord_flip() +
  theme_info +
  theme(
    legend.position = "none",
    axis.title.y = element_blank(),
    axis.title.x = element_text(size = 13),
    axis.text    = element_text(size = 8)
  )

# Create final figure
plot_grid(
  clust_umap, ova_boxes,
  rel_widths = c(0.44, 0.56),
  nrow  = 1,
  align = "h",
  axis  = "trbl"
)

```

<br>

### {.tabset .tabset-pills}

Marker genes were identified by comparing all clusters. `r marker_desc` `r GO_desc`
```{r "{{.y}} Cluster markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Create figures
fig_umap <- clust_umap +
  guides(color = col_guide) +
  theme(
    legend.position = "top",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 10)
  )

create_marker_panel_v1(
  input_sobj   = sobj,
  input_cols   = clust_cols,
  input_umap   = fig_umap,
  clust_column = "cluster_id"
)

```

### Subtype Cluster Markers

A UMAP projection colored by cell cluster is shown on the left and ovalbumin signal for each cluster is shown on the right. Clusters are grouped by subtype.
```{r "{{.y}} Subtype cluster OVA", fig.width = 9, fig.height = 4.2}

# Plot OVA counts
ova_data <- sobj %>%
  FetchData(c("adt_ovalbumin", "cell_type", "cluster_id")) %>%
  as_tibble(rownames = "cell_id") %>%
  group_by(cell_type) %>%
  mutate(type_med = median(adt_ovalbumin)) %>%
  group_by(cluster_id) %>%
  mutate(clust_med = median(adt_ovalbumin)) %>%
  ungroup() %>%
  arrange(type_med, clust_med) %>%
  mutate(cluster_id = fct_inorder(cluster_id))
  
ova_boxes <- ova_data %>%
  mutate(adt_ovalbumin = adt_ovalbumin + 0.01) %>%
  ggplot(aes(cluster_id, adt_ovalbumin, color = cluster_id)) +
  geom_quasirandom(size = 0.25 * mtplyr_2) +
  scale_color_manual(values = clust_cols) +
  scale_y_log10() +
  coord_flip() +
  theme_info +
  theme(
    strip.text = element_blank(),
    legend.position = "none",
    axis.title.y = element_blank(),
    axis.title.x = element_text(size = 13),
    axis.text    = element_text(size = 10)
  )

# Create final figure
plot_grid(
  clust_umap, ova_boxes,
  rel_widths = c(0.44, 0.56),
  nrow  = 1,
  align = "h",
  axis  = "trbl"
)

```

<br>

### {.tabset .tabset-pills}

Marker genes were identified by splitting the cells based on subtype and comparing the clusters for each subtype. `r marker_desc` `r GO_desc`
```{r "{{.y}} Subtype cluster markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Find cluster markers separately for each subtype
fig_markers <- cell_types %>%
  map(~ {
    find_group_markers(
      input_grp  = .x,
      input_sobj = sobj,
      grp_col    = "cell_type",
      clust_col  = "cluster_id"
    )
  }) %>%
  compact() %>%
  bind_rows()

# Create figures
create_marker_panel_v2(
  input_sobj    = sobj,
  input_markers = fig_markers,
  input_cols    = clust_cols,
  grp_column    = "cell_type",
  clust_column  = "cluster_id"
)

```

### Ovalbumin Markers

Cells were divided into groups based on ovalbumin signal. The ovalbumin groups are shown on the left and ovalbumin signal is shown on the right. The proportion of subtypes and cell clusters that make up each group is shown at the bottom.
```{r "{{.y}} OVA cluster summary", fig.width = 8.5, fig.height = 9}

# Determine number of OVA groups
n_grps <- floor(ncol(sobj) / 100)

n_grps <- case_when(
  n_grps > 6 ~ 6,
  n_grps < 2 ~ 2,
  TRUE ~ n_grps
)

# Set OVA colors
ova_cols <- c(
  "#6a51a3",  # purple 1
  "#8073ac",  # purple 2
  "#9e9ac8",  # purple 3
  "#ECBA96",  # red 3
  "#fb6a4a",  # red 2
  "#e31a1c"   # red 1
)

ova_cols <- ova_cols[1:n_grps]
names(ova_cols) <- as.character(1:n_grps)

# Divide cells based on OVA signal
sobj <- sobj %>%
  FetchData("adt_ovalbumin") %>%
  rownames_to_column("cell_id") %>%
  mutate(
    ova_group = ntile(adt_ovalbumin, n_grps),
    ova_group = as.character(ova_group)
  ) %>%
  column_to_rownames("cell_id") %>%
  AddMetaData(
    object = sobj,
    metadata = .
  )

# OVA group UMAP
ova_grp_umap <- sobj %>%
  plot_features(
    feature    = "ova_group",
    pt_size    = 0.1 * mtplyr_3,
    pt_outline = 0.4,
    plot_cols  = ova_cols
  ) +
  guides(color = col_guide) +
  ggtitle("Ovalbumin Group") +
  blank_theme +
  theme(
    plot.title      = element_text(size = 13),
    legend.position = "right",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8)
  )

# OVA UMAP
ova_umap <- sobj %>%
  plot_features(
    feature    = "adt_ovalbumin",
    pt_size    = 0.1 * mtplyr_3,
    pt_outline = 0.4,
    plot_cols  = c("#fafafa", type_cols[2])
  ) +
  ggtitle("adt_ovalbumin") +
  blank_theme +
  theme(
    plot.title        = element_text(size = 13),
    legend.position   = "bottom",
    legend.key.height = unit(0.1, "cm"),
    legend.key.width  = unit(0.3, "cm"),
    legend.title      = element_blank(),
    legend.text       = element_text(size = 8)
  )

# OVA hist
ova_hist <- sobj@meta.data %>%
  mutate(adt_ovalbumin = adt_ovalbumin + 0.01) %>%
  ggplot(aes(adt_ovalbumin, fill = ova_group)) +
  geom_histogram(position = "identity", bins = 30) +
  scale_fill_manual(values = ova_cols) +
  scale_x_log10() +
  guides(fill = guide_legend(ncol = 1)) +
  labs(title = "Ovalbumin Group", y = "Cell Count") +
  theme_info +
  theme(
    plot.title      = element_text(size = 13),
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8),
    axis.title      = element_text(size = 13),
    axis.text       = element_text(size = 8),
  )

# OVA subtype bar graphs
type_bar <- sobj %>%
  plot_cell_count(
    group_id = "ova_group",
    split_id = NULL,
    fill_id = "cell_type",
    plot_colors = type_cols
  ) +
  labs(x = "ova_group") +
  guides(fill = guide_legend(ncol = 1)) +
  ggtitle("Subtype") +
  theme(
    plot.title      = element_text(size = 13),
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8),
    axis.title.x    = element_text(size = 13),
    axis.text       = element_text(size = 8),
  )

# OVA cluster bar graphs
clust_bar <- sobj %>%
  plot_cell_count(
    group_id = "ova_group",
    split_id = NULL,
    fill_id = "cluster_id",
    plot_colors = clust_cols
  ) +
  labs(x = "ova_group") +
  guides(fill = guide_legend(nrow = 10)) +
  ggtitle("Cluster") +
  theme(
    plot.title      = element_text(size = 13),
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8),
    axis.title      = element_text(size = 13),
    axis.text.x     = element_text(size = 8),
    axis.title.y    = element_blank(),
    axis.text.y     = element_blank(),
    axis.ticks.y    = element_blank(),
    axis.line.y     = element_blank()
  )

blank <- ggplot() +
  geom_blank() +
  theme_void()

# Create final figure
umaps <- plot_grid(
  ova_grp_umap, ova_umap, blank,
  rel_widths = c(0.33, 0.33, 0.05),
  nrow  = 1,
  align = "hv",
  axis  = "trbl"
)

bars <- plot_grid(
  ova_hist, type_bar, clust_bar,
  rel_widths = c(0.5, 0.365, 0.305),
  nrow  = 1,
  align = "h",
  axis  = "tb"
)

plot_grid(
  umaps, bars,
  rel_heights = c(0.45, 0.55),
  nrow = 2,
  align = "vh",
  axis  = "trbl"
)

```

<br>

### {.tabset .tabset-pills}

Marker genes were identified by comparing all ovalbumin groups. `r marker_desc` `r GO_desc`
```{r "{{.y}} OVA cluster markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Create figures
create_marker_panel_v1(
  input_sobj    = sobj,
  input_cols    = ova_cols,
  clust_column  = "ova_group"
)

```

### Subtype Ovalbumin Markers

Each subtype was divided into groups based on ovalbumin signal. A UMAP projection colored by the subtype ovalbumin groups is shown on the left and ovalbumin signal is shown on the right.
```{r "{{.y}} Subtype OVA groups summary", fig.width = 8.5, fig.height = 6}

# Divide subtypes based on OVA signal
sobj <- sobj %>%
  FetchData(c("cell_type", "adt_ovalbumin")) %>%
  rownames_to_column("cell_id") %>%
  group_by(cell_type) %>%
  mutate(
    n_grps       = if_else(n() < 300, 2, 3),
    ova_group    = ntile(adt_ovalbumin, n_grps),
    ova_group    = as.character(ova_group),
    type_ova_grp = str_c(ova_group, "-", cell_type)
  ) %>%
  ungroup() %>%
  column_to_rownames("cell_id") %>%
  AddMetaData(
    object = sobj,
    metadata = .
  )

# Set OVA colors
type_ova_cols <- theme_cols_threes[1:length(cell_types)] %>%
  reduce(c)

names(type_ova_cols) <- cell_types %>%
  map(~ str_c(1:3, "-", .x)) %>%
  reduce(c)

# OVA group UMAP
ova_grp_umap <- sobj %>%
  plot_features(
    feature     = "type_ova_grp",
    pt_size     = 0.1 * mtplyr_3,
    pt_outline  = 0.4,
    plot_cols   = type_ova_cols,
    feat_levels = names(type_ova_cols)
  ) +
  guides(color = guide_legend(override.aes = list(size = 3.5, shape = 16), ncol = 2)) +
  blank_theme +
  theme(
    legend.position = "bottom",
    legend.title    = element_blank(),
    legend.text     = element_text(size = 8)
  )

# OVA boxes
ova_data <- sobj %>%
  FetchData(c("adt_ovalbumin", "cell_type", "type_ova_grp")) %>%
  as_tibble(rownames = "cell_id") %>%
  group_by(cell_type) %>%
  mutate(type_med = median(adt_ovalbumin)) %>%
  group_by(type_ova_grp) %>%
  mutate(clust_med = median(adt_ovalbumin)) %>%
  ungroup() %>%
  arrange(type_med, clust_med) %>%
  mutate(type_ova_grp = fct_inorder(type_ova_grp))
  
ova_boxes <- ova_data %>%
  mutate(adt_ovalbumin = adt_ovalbumin + 0.01) %>%
  ggplot(aes(cell_type, adt_ovalbumin, color = type_ova_grp)) +
  geom_quasirandom(size = 0.25 * mtplyr_2) +
  scale_color_manual(values = type_ova_cols) +
  scale_y_log10() +
  coord_flip() +
  theme_info +
  theme(
    legend.position = "none",
    axis.title.y    = element_blank(),
    axis.title.x    = element_text(size = 13),
    axis.text       = element_text(size = 10)
  )

# Create final figure
plot_grid(
  ova_grp_umap, ova_boxes,
  rel_widths = c(0.44, 0.56),
  nrow  = 1,
  align = "h",
  axis  = "trbl"
)

```

### {.tabset .tabset-pills}

Marker genes were identified by splitting the cells based on subtype and comparing the ovalbumin groups within the subtype. `r marker_desc` `r GO_desc`
```{r "{{.y}} Subtype OVA markers", fig.width = 8.5, fig.height = 14, results = "asis"}

# Find OVA markers separately for each subtype
fig_markers <- cell_types %>%
  map(~ {
    find_group_markers(
      input_grp  = .x,
      input_sobj = sobj,
      grp_col    = "cell_type",
      clust_col  = "type_ova_grp"
    )
  }) %>%
  compact() %>%
  bind_rows()

# Create figures
create_marker_panel_v2(
  input_sobj       = sobj,
  input_markers    = fig_markers,
  input_cols       = type_ova_cols,
  grp_column       = "cell_type",
  clust_column     = "type_ova_grp",
  n_boxes          = 34,
  all_boxes        = T,
  panel.background = element_rect(color = "#fafafa", fill = "#fafafa"),
  panel.spacing.y  = unit(0, "cm"),
  panel.spacing.x  = unit(0.2, "cm"),
  strip.text       = element_text(size = 10, angle = 90, hjust = 0)
)

```


```{r, eval = F, echo = F}

x <- GO_list %>%
  compact() %>%
  bind_rows()

top_terms <- x %>%
  filter(source == "GO:BP") %>%
  group_by(cluster) %>%
  top_n(10, desc(p_value)) %>%
  ungroup() %>%
  pull(term_name) %>%
  unique()

# all_terms <- unique(x$term_name)
# all_terms <- all_terms[!all_terms %in% top_terms]

GO_ova_data <- top_terms %>%
  map(~ {
    clusts <- x %>%
      filter(
        source == "GO:BP",
        term_name == .x
      ) %>%
      pull(cluster)
    
    ova_counts <- sobj %>%
      FetchData(c("cluster_id", "adt_ovalbumin")) %>%
      as_tibble(rownames = "cell_id") %>%
      filter(cluster_id %in% clusts) %>%
      mutate(term_name = .x)
  }) %>%
  compact() %>%
  bind_rows()

GO_box_data <- GO_ova_data %>%
  group_by(term_name) %>%
  mutate(
    med_ova = median(adt_ovalbumin),
    up_qt = quantile(adt_ovalbumin)[4]
  ) %>%
  ungroup() %>%
  mutate(term_name = fct_reorder(term_name, adt_ovalbumin, median))

y_max <- max(GO_box_data$up_qt) * 1.1

GO_box_data %>%
  ggplot(aes(term_name, adt_ovalbumin, fill = med_ova)) +
  geom_boxplot(size = 0.1, color = "white", outlier.shape = NA) +
  coord_flip(ylim = c(0, y_max)) +
  theme_info +
  theme(
    legend.position = "none",
    axis.title.y    = element_blank(),
    axis.text       = element_text(size = 8)
  )

  # arrange(med_ova) %>%
  # tail(10) %>%
  # mutate(term_name = fct_inorder(term_name)) %>%

```

```{r "{{.y}} GO TEST", eval = F, echo = F}

# Find markers
Idents(sobj) <- sobj %>%
  FetchData("ova_group")

test_markers <- find_markers(sobj)

# Filter markers for OVA group
grp_markers <- test_markers %>%
  filter(cluster == 6) %>%
  pull(gene)

# Run gprofiler2
test_GO <- grp_markers %>%
  run_gprofiler(
    genome        = "GRCm",
    ordered_query = T,
    evcodes       = T
  )

# Extract intersecting genes
antigen_genes <- test_GO[[1, "intersection"]] %>%
  str_split(",") %>%
  .[[1]]

# Add module score
test_so <- sobj %>%
  AddModuleScore(
    assay    = "RNA",
    features = list(antigen_genes),
    name     = "antigen",
    ctrl     = 50
  )

# UMAP showing antigen module score
test_so %>%
  plot_features(
    feature   = "antigen1",
    pt_size   = 1
  ) +
  guides(color = col_guide) +
  blank_theme

# UMAP showing OVA signal
test_so %>%
  plot_features(
    feature   = "adt_ovalbumin",
    pt_size   = 1
  ) +
  guides(color = col_guide) +
  blank_theme

# UMAP showing OVA group
test_so %>%
  plot_features(
    feature   = "ova_group",
    plot_cols = theme_cols,
    pt_size   = 1,
    feat_levels = c("6", "5", "4", "3", "2", "1")
  ) +
  guides(color = col_guide) +
  blank_theme



# mart_obj <- useMart(
#   biomart = "ensembl", 
#   dataset = "mmusculus_gene_ensembl",
#   verbose = T
# )

# antigen_genes <- getBM(
#   attributes = c("ensembl_gene_id","external_gene_name"), 
#   filters    = "go",
#   values     = "GO:0002495",
#   mart       = mart_obj
# )

```



























